### ISSUE TITLE
Replace main args on jackin

### ISSUE THREAD

**Issue by Cyrik** [BetterThanTomorrow/calva#1682]
When jacking into a project that has main-args in one of the aliases Calva gives a warning that this is going to break. This is a good start, but we could help more by changing the main-args in that case. Trying to jack-in to clojure-lsp as an example: 
cider
`clojure -Sdeps '{:deps {nrepl/nrepl {:mvn/version "0.9.0"} cider/cider-nrepl {:mvn/version "0.28.3"} refactor-nrepl/refactor-nrepl {:mvn/version "3.5.2"}} :aliases {:cider/nrepl {:main-opts ["-m" "nrepl.cmdline" "--middleware" "[refactor-nrepl.middleware/wrap-refactor,cider.nrepl/cider-middleware]"]}}}' -Mdev:test:cider/nrepl`
Calva
`clojure -Sdeps '{:deps {nrepl/nrepl {:mvn/version,"0.9.0"},cider/cider-nrepl {:mvn/version,"0.27.4"}}}' -M:dev:test:debug`

This is probably more intuitive for beginners and would allow us to set dev/test selected by default, which most projects use anyway.

**Comment by PEZ** [comment]
I like this!

**Comment by seancorfield** [comment]
What about people who use an alias to start nREPL with custom code? Whatever solution is picked has to be able to NOT override the user's main opts as well.

If Calva did this by default, my entire workflow would be broken since I always use a `:dev/repl` alias to start things up - I don't want Calva's main opts.

**Comment by Cyrik** [comment]
@seancorfield at that point, wouldn't it be simpler to start your own repl and auto-connect to it from calva instead of starting it from calva?
Note that I don't want to disregard/disable your use case, just trying to figure out how we can make both versions work.

**Comment by seancorfield** [comment]
The custom jack-in stuff has allowed me to switch most of my projects to my dot-clojure and `:dev/repl` setup, so Calva can manage starting and connecting to my REPL. Prior to that, I used to start my own REPLs manually because Calva couldn't do what I needed. I hadn't realized how convenient that would be until Calva could actually do it.

With a very minor change to one other project, it now means I can open up pretty much any project and jack-in -- with Calva using my nicely customized REPL -- so it seems like a big step backward to essentially break/remove that workflow option.

I've suggested a couple of times offering an additional checkbox in the alias dropdown that, if checked, would suppress the `:main-opts` of the alias (and have Calva provide its regular main options). That checkbox could even by checked by default.

That way, the default for everyone would be what is suggested here: aliases with `:main-opts` would have their `:main-opts` ignored essentially because Calva could provide a unique alias with what it needed for `:main-opts` which would then override any selected alias `:main-opts`. But people who have an alias that would start a REPL could **uncheck** that box next to their selected alias and Calva would then **not** add its own main options.

If multiple aliases have `:main-opts`, Calva should put the unchecked alias last (so its `:main-opts` would "win"). If a use unchecks multiple aliases' "Suppress :main-opts" checkbox, that would be undefined behavior (Calva could put those aliases "last" in alphabetical order).

So Calva would select "all checked aliases except those with unchecked suppression" followed by "checked aliases with unchecked suppression", and if the latter set was empty, it would add its own alias with `:main-opts` and append that to the list.

**Comment by PEZ** [comment]
With @seancorfield's feedback I lean towards that the current behaviour is better than always providing Calva's own main-opts.

I hesitate to add more checkboxes to the list, but I think it is worth considering.

How about a separate setting: `calva.depsEdnJackInMainOpts` which defaults to the options Calva needs. Calva will use that regardless if main-opts aliases are selected, as the OP suggests. Setting it to `null` would stop Calva from adding it.

**Comment by seancorfield** [comment]
I don't think you can rely on settings like that, even within a workspace. I have user settings to always show `:portal` and `:dev/repl` aliases which come from my dot-clojure  files, which I can choose to use with _any_ project I pull down off the web.

I may want to use the standard jack-in some times and my `:dev/repl` `:main-opts` other times. Whatever solution is taken to this needs to be on a per-jack-in basis in terms of selecting which aliases to use or not.

That's what drives me to something within the alias selection drop-down.

Perhaps a "simple" option here might be:
* change Calva to supply its `:main-opts` in a unique alias that it adds via `-Sdeps`
* always add, as the last entry, that alias to the drop-down, after user-configured options and have it checked by default

The default behavior is then:
* jack-in, for `deps.edn` project
* show the alias list (change nothing)
* jack-in uses the Calva alias to engage Calva's `:main-opts`

My workflow then becomes:
* jack-in, for `deps.edn` project
* show the alias list and choose whatever aliases I want
* uncheck the Calva alias _if I selected an alias that will start a REPL_
* jack-in uses the selected aliases

That adds no new checkbox to the list -- just a final line item that is Calva's alias and checked by default -- and would default to Calva's main options being used per @Cyrik request here.

It would also address the situation where a project's `:test` alias contained `:main-opts` to run tests (per Cognitect's `test-runner` README) and still have Calva's main options come into play -- thus addressing a shortfall with the status quo.

I would also request, if you go down this route, to change the order of the alias list so that it is:
* project aliases, alphabetical
* extra aliases from settings (user + workspace), alphabetical
* Calva alias

The reason for wanting the project and extra aliases separated is that, currently, `:dev/repl` from my (user) settings sorts before `:test` from a project `deps.edn` and if both have `:main-opts`, `:test` wins which is pretty much always going to be wrong. My only recourse here is to rename my REPL-starter to something like `:zzz_dev/repl` to force it to sort last (and win). Ugh!

**Comment by PEZ** [comment]
Thanks! It makes sense. The only culprit is that it introduces a point where the user can go wrong, by unselecting Calva's main-opts. We could mitigate that by only presenting this option when there are competing main-opts. Problem there is that we don't always know this. Maybe it's time we start to read the user deps config...
